core/drivers/crypto/caam/mp/caam_mp.c

```diff
@@ -0,0 +1,404 @@
+	caam_desc_add_ptr(desc, sign_d.sgtbuf.paddr);
+	caam_desc_add_word(desc, msg_input.sgtbuf.length);
+	caam_desc_add_word(desc, MPSIGN_OP);
+
+	desclen = caam_desc_get_len(desc);
+	caam_desc_update_hdr(desc, DESC_HEADER_IDX(desclen, desclen - 1));
+
+	MP_DUMPDESC(desc);
+
+	caam_dmaobj_cache_push(&msg_input);
+	caam_dmaobj_cache_push(&sign_c);
+
+	jobctx.desc = desc;
+	retstatus = caam_jr_enqueue(&jobctx, NULL);
+
+	if (retstatus == CAAM_NO_ERROR) {
+		sign_c.orig.length = 2 * mp_privdata.sec_size;
+		*sig_size = caam_dmaobj_copy_to_orig(&sign_c);
+
+		MP_DUMPBUF("MP Signature", sdata->signature.data,
+			   sdata->signature.length);
+
+		ret = caam_status_to_tee_result(retstatus);
+	} else {
+		MP_TRACE("CAAM Status 0x%08" PRIx32, jobctx.status);
+		ret = job_status_to_tee_result(jobctx.status);
+	}
+
+exit_mpsign:
+	caam_free(aligned_msg);
+	caam_free_buf(&hash);
+	caam_free_desc(&desc);
+	caam_dmaobj_free(&msg_input);
+	caam_dmaobj_free(&sign_c);
+	caam_dmaobj_free(&sign_d);
+
+	return ret;
+}
+
+enum caam_status caam_mp_init(vaddr_t ctrl_addr)
+{
+	enum caam_status retstatus = CAAM_FAILURE;
+	uint8_t curve = 0;
+	uint8_t hash_limit = 0;
+
+	struct caambuf msg_mpmr = {
+		.data = (uint8_t *)g_mpmr_data,
+		.length = strlen(g_mpmr_data)
+	};
+
+	mp_privdata.ctrl_addr = ctrl_addr;
+	mp_privdata.mp_status = CAAM_NOT_INIT;
+
+	curve = caam_hal_ctrl_get_mpcurve(ctrl_addr);
+
+	if (curve == UINT8_MAX) {
+		mp_privdata.mp_status = CAAM_NOT_SUPPORTED;
+		return mp_privdata.mp_status;
+	}
+
+	if (caam_hal_ctrl_is_mp_set(ctrl_addr)) {
+		mp_privdata.mp_status = CAAM_NO_ERROR;
+		return CAAM_NO_ERROR;
+	}
+
+	if (!curve) {
+		/* Get the device HASH Limit to select the MP Curve */
+		hash_limit = caam_hal_ctrl_hash_limit(ctrl_addr);
+
+		switch (hash_limit) {
+		case TEE_MAIN_ALGO_SHA256:
+			mp_privdata.curve = PDB_MP_CSEL_P256;
+			mp_privdata.sec_size = 32;
+			break;
+		case TEE_MAIN_ALGO_SHA512:
+			mp_privdata.curve = PDB_MP_CSEL_P521;
+			mp_privdata.sec_size = 66;
+			break;
+		default:
+			MP_TRACE("This curve doesn't exist");
+			return CAAM_FAILURE;
+		}
+
+		MP_TRACE("MP Private key has not been generated");
+		retstatus = do_mppriv_gen(g_passphrase, strlen(g_passphrase));
+
+		if (retstatus != CAAM_NO_ERROR) {
+			MP_TRACE("do_mppriv_gen failed!");
+			return retstatus;
+		}
+	} else {
+		/* MP Curve is already programmed. Set the right key size */
+		mp_privdata.curve = curve;
+
+		switch (curve) {
+		case PDB_MP_CSEL_P256:
+			mp_privdata.sec_size = 32;
+			break;
+		case PDB_MP_CSEL_P521:
+			mp_privdata.sec_size = 66;
+			break;
+		default:
+			MP_TRACE("This curve is not supported");
+			return retstatus;
+		}
+	}
+
+	/* Fill the MPMR content then lock it */
+	caam_hal_ctrl_fill_mpmr(ctrl_addr, &msg_mpmr);
+
+	mp_privdata.mp_status = CAAM_NO_ERROR;
+
+	return CAAM_NO_ERROR;
+}
+
+void caam_mp_resume(uint32_t pm_hint)
+{
+	if (pm_hint == PM_HINT_CONTEXT_STATE)
+		caam_mp_init(mp_privdata.ctrl_addr);
```

Comment:
should test the return value

Reply:
I will print a message in case of a failure